Clean up the code for validation

nicholas-david opened the pull request:
Create a more streamlined coding process for the validation scripts. 

- [x] Create configuration files for edge cities
- [x] Create configuration files for poi cities
- [x] Understand how to protect data
- [x] Reformat edge_validation.py to pull from the configuration files
- [x] Reformat edge_validation.py to define specific functions for future notebooks
- [x] Clean up jupyter notebooks in edge folder
- [ ] Fix bugs in the in edge validation module
- [ ] Reformat poi_validation.py to pull from the configuration files
- [ ] Reformat poi_validation.py to define specific functions for future notebooks
- [ ] Clean up jupyter notebooks in poi folder

nicholas-david committed:
Create data folders for poi and validation

nicholas-david commented:
@gboeing Hey Geoff, how do I prevent the data from the collaborators being committed to the repo? I have it on my machine and in the new data folders that I created in validation. Thanks!

gboeing commented:
You can edit the [.gitignore](https://github.com/gboeing/global-indicators/blob/master/.gitignore) file to tell git to ignore those data files (or their containing folders).

nicholas-david committed:
Update to the gitignore

nicholas-david committed:
delete readme file in edge data foldler

nicholas-david committed:
merge

nicholas-david committed:
add config files for the different sources of officail data

nicholas-david committed:
merge from upstream

nicholas-david committed:
adding configuration filepaths to the python script in the validation edge directory

nicholas-david committed:
merging from upstream

nicholas-david committed:
Update edge_validation.py

Defines the different functions

nicholas-david committed:
Update edge_validation.py

Refine the document and make it compatible with jupyter notebook

nicholas-david committed:
Update edge_validation.py

Some final adjustments to clean everything up. Removed section that saves images.

nicholas-david committed:
Update edge_validation.py

Trying to fix bugs so that data can be loaded using this module

nicholas-david commented:
@shiqin-liu Hey, can you look at the module 'edge_validation.py' and see if you can help me understand how to import the data from configuration files. Thanks!

gboeing commented on validation/edge/edge_validation.py:
```diff
@@ -13,152 +13,144 @@
-    print("-----------------------")
+
+
+if __name__ == '__main__':
```
To make this .py file usable as a module rather than a script to be run at the command line, you'll want to remove this line and you'll want to wrap all the code inside this file into functions. That way, everything in here can be access from the notebook like:

```python
import edge_validation as ev
total_length = ev.get_total_length(straightforward_variable_name)
# etc
```

gboeing commented:
@nicholas-david if we merge your validation cleanups in soon, I can do some follow up code cleanup and refactoring.

nicholas-david committed:
Update edge_validation.py

Bug fies

nicholas-david committed:
Simplify Config Files

nicholas-david committed:
Update edge_validation-Belfast.ipynb

nicholas-david committed:
Update edge so that notebook runs

nicholas-david committed:
Update edge_validation-Belfast.ipynb

nicholas-david committed:
update branch from master

nicholas-david committed:
Further Updates

nicholas-david commented:
@gboeing I am still getting an error that I am unsure how to deal with. The module is getting stuck on line 51. I am getting the following error:

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf3 in position 31: invalid continuation byte

Please let me know how I can fix this issue. Thanks!

gboeing commented:
What file are you opening when you get that error? It sounds like an encoding conflict. My guess is that this file is not encoded in utf-8, which is what geopandas expects by default.

nicholas-david commented:
@gboeing 
I am trying to load the following:
gdf_study_area = gpd.read_file(osm_buffer, layer='urban_study_region')
where OSM buffer is the following file:
belfast_gb_2019_1600m_buffer.gpkg

gboeing commented:
I'm unable to reproduce that error. I just ran this code:

```python
import geopandas as gpd
osm_buffer = 'belfast_gb_2019_1600m_buffer.gpkg'
gdf_study_area = gpd.read_file(osm_buffer, layer='urban_study_region')
gdf_study_area.shape #(1, 2)
```

It works successfully... try re-downloading the geopackage file again in case it got corrupted?

nicholas-david commented:
@gboeing 
When I run that code, I also am able to run without issue. When I run the line of code in my module, however, I still get the error described above.

gboeing commented:
I can clone your fork and test directly. Can you provide a complete reproducible code snippet for me to run using your fork's validation branch?

nicholas-david commented:
@gboeing 
What is the best way for me to do that?

gboeing commented:
You can just paste in a set of steps for me to follow or lines of code for me to run here. Just provide every step so I can reproduce precisely how you generate the error.

nicholas-david commented:
In the module:

```
#import modules
import osmnx as ox
import networkx as nx
import geopandas as gpd
import pandas as pd
import pandana as pdna
import numpy as np
from scipy.stats import zscore
import gc
import os
import csv
import matplotlib.pyplot as plt
from shapely.geometry import Polygon, mapping
import json
import sys
from os.path import join as pjoin

# For 'filename', insert belfast.json, hong_kong.json, or olomouc.json
def load_data(filename, city):
    # Get the work directory
    dirname = os.path.abspath('')
    # Define the path for the city-specifc configuration file
    jsonFile = 'configuration/' + filename
    jsonPath = os.path.join(dirname, jsonFile)
    # Use the defined paths to run the config file
    try:
        with open(jsonPath) as json_file:
            config = json.load(json_file)
    except Exception as e:
        print('Failed to read json file.')
        print(e)

    # Read the filepaths so data can be accessed
    graphml_path = os.path.join(dirname, 'data', city, config['graphml_path'])
    osm_buffer_path = os.path.join(dirname, 'data', city, config['osm_buffer'])
    gdf_official_path = os.path.join(dirname, 'data', city, config ['gdf_official'])

    # Open the files from the filepaths
    graphml = open(graphml_path)
    osm_buffer = open(osm_buffer_path)
    gdf_official = open(gdf_official_path)

    return (graphml, osm_buffer, gdf_official)

def refine_data(graphml, osm_buffer, gdf_official):
    # Extract specific layers from the data
    G = ox.load_graphml(graphml)
    G_undirected = ox.get_undirected(G)
    gdf_osm = ox.utils_graph.graph_to_gdfs(G_undirected, nodes=False, edges=True)
    gdf_study_area = gpd.read_file(osm_buffer, layer='urban_study_region')
    # Project the data to a common crs
    gdf_osm = gdf_osm.to_crs(gdf_official.crs)
    gdf_study_area = gdf_study_area.to_crs(gdf_official.crs)

    return (gdf_osm, gdf_study_area)
```

nicholas-david commented:
In the notebook:
```
import json
import edge_validation as ev
```

```
with open('configuration/belfast.json') as f:
    config = json.load(f)
```

`(graphml, osm_buffer, gdf_official) = ev.load_data('belfast.json', 'Belfast')
`
`(gdf_osm, gdf_study_area) = ev.refine_data(graphml, osm_buffer, gdf_official)`

nicholas-david commented:
@gboeing does this work?

gboeing commented:
where do i find the belfast.json file?

gboeing commented:
nvm, i pulled it from your branch

gboeing commented on validation/edge/edge_validation.py:
```diff
@@ -11,154 +11,122 @@
+
+    # Open the files from the filepaths
+    graphml = open(graphml_path)
+    osm_buffer = open(osm_buffer_path)
```
@nicholas-david the problem is here. You're opening the file itself, then taking the open file and asking geopandas to open it. This should just return `osm_buffer_path`, that's all geopandas needs to open to the file.

nicholas-david replied:
Thank you for clearing that up. The reason that I opened the file using those lines is that I was getting an error that says strings do not have the attribute .crs. when trying to change the crs to standardized them between data sources. 

Is there a way to fix that issue?

gboeing replied:
I made that code edit that I suggested, then re-ran the code snippet you provided earlier, and it all seemed to run fine.

gboeing commented:
@nicholas-david if you are down to just minor bug-squashing, it might be a good time to merge the PR. I can take a pass on Friday to do another round of code cleanup and bug fixes prior to our Monday morning discussion.

nicholas-david committed:
Update Belfast

nicholas-david committed:
Rename files to use 'destination'

Help make nomenclature consistent hroughout project.

nicholas-david committed:
Update belfast notebook

Include markdown comments for each cell

nicholas-david commented:
@gboeing Okay! I think that it is ready to merge

gboeing commented:
@nicholas-david thanks! Looking good. I'll aim to take one more pass through it tomorrow.
